src/connection.c

```diff
@@ -330,3 +328,24 @@ ConnectionGetOptionValue(PGconn *connection, char *optionKeyword)
 
 	return optionValue;
 }
+
+ShardId DoForAllShards(List* shardConnections, ShardAction action, void* arg)
+{
+	ListCell *listCell;
+	int i = 0;
+
+	foreach (listCell, shardConnections)
+	{
+		ShardConnections* shardConn = (ShardConnections*)lfirst(listCell);
+		bool allOk = true;
+		for (i = 0; i < shardConn->replicaCount; i++) 
+		{						
+			allOk &= shardConn->status[i] = action(shardConn->shardId, shardConn->conn[i], arg, shardConn->status[i]);
+		}
+		if (!allOk) 
```

Comment:
DoForAllShards terminates after a single shard fails to perform commit/abort, but this would leave a lot of prepared transactions for all the other shards. I think it would be preferable to produce a warning (e.g. "WARNING: transaction data_192 failed to commit on 192.168.2.3:5432") and continue committing/aborting the other ones, such that the user can step in and recover only those transactions that failed. When a prepare fails we do want to stop immediately and roll back. Having a generic DoForAllShards might not be the right answer here.

Given that the loop is pretty simple now, you could move it into the CopyPrepareTransaction, CopyAbortTransaction, and CopyEndTransaction functions instead of having those act on a single shard and produce the appropriate warnings there.